### PR TITLE
Nicks branch

### PR DESCRIPTION
**Pull request by NickDeas**
Conditional saving as flattened json
Support for user and referenced tweet expansions
Updated readme and default value reconfiguration

### CODE REVIEW

**Comment by patrick-lee-warren** [comment]
I tested this branch, with the code 
`pull_twitter.py -cf '/home/pwarren/pull_twitter/configs/plw_config.yaml' search -q 'Trump OR pelosi'  -st '2017-07-14' -et '2017-07-16' -mr 100`

and got the error:
`Pulling tweet results using 'Trump OR pelosi' search query.
Saving tweets to /zfs/disinfo/pull_twitter_test/search/search_2021-10-19 12.24.16/data_%s.csv
Failed to pull tweets for query. Error:  data type 'Float64' not understood`

**Comment by patrick-lee-warren** [comment]
> I tested this branch, with the code `pull_twitter.py -cf '/home/pwarren/pull_twitter/configs/plw_config.yaml' search -q 'Trump OR pelosi' -st '2017-07-14' -et '2017-07-16' -mr 100`
> 
> and got the error: `Pulling tweet results using 'Trump OR pelosi' search query. Saving tweets to /zfs/disinfo/pull_twitter_test/search/search_2021-10-19 12.24.16/data_%s.csv Failed to pull tweets for query. Error: data type 'Float64' not understood`

Mystery solved! You need to be running Python 3.8, so the read.me probably needs to updated accordingly, since it currently says you can run in 3.7.

File: utils/timeline.py [hunk 1]
```diff
@@ -86,14 +90,84 @@ def pull(self,
...
+            inc_tweets: List[dict] = response.includes['tweets'] if 'tweets' in response.includes.keys() else None
+            inc_users: List[dict] = response.includes['users'] if 'users' in response.includes.keys() else None
+            # inc_media: List[dict] = response.includes['media'] if 'media' in response.includes.keys() else None
             if tweets:
+                
+                # Expansions parsing
+                if inc_tweets:
+                    ref_tweets = [twalc.Tweet(**tw).to_dict() for tw in inc_tweets]
+                if inc_users:
+                    authors = [twalc.User(**us).to_dict() for us in inc_users]
```
Comment (dhudsmith) [inline utils/timeline.py:103]: rename "users"

File: utils/timeline.py [hunk 2]
```diff
@@ -86,14 +90,84 @@ def pull(self,
...
+                if has_refs:
+                    links = Timeline.__parse_tweet_links(tweets)
+
+                # Original Tweets Parsing
                 tweets = [twalc.Tweet(**tw).to_dict() for tw in tweets]
+                
+
+                # Expansions dataframes
+                if inc_tweets:
+                    df_refs   = pd.concat([df_refs, pd.DataFrame(ref_tweets)], axis = 0) if df_refs is not None else pd.DataFrame(ref_tweets)
```
Comment (dhudsmith) [inline utils/timeline.py:115]: is there any harm in using `dtype=str` inside pd.DataFrame to force all fields to be treated as strings? That would sidestep the need for the fix floats function (and I think the dependency on python version >=3.8). I guess the public metrics are a little inefficient as strings but that doesn't seem very important.

File: utils/tweet_search.py [hunk 3]
```diff
@@ -28,7 +28,7 @@ def __init__(self,
 		self.client: Client = tweepy_client
 		self.query_params = query_params
 
-	def pull(self, query:str, output_dir: str, 
+	def pull(self, query:str, output_dir: str, save_format: str = 'csv',
```
Comment (dhudsmith) [inline utils/tweet_search.py:31]: should args be class members?

File: utils/tweet_search.py [hunk 4]
```diff
@@ -130,4 +203,29 @@ def search_tweets(self, query: str,
...
+		#Select all float columns and Float64 columns (coerced after concatenating nullable values)
+		float_cols = df.select_dtypes(include=[float, "Float64"])
+
+		# Convert float columns to nullable Int64
+		for col in float_cols:
+			df[col] = df[col].astype('Int64')
+
+		return df
+
+	@staticmethod
```
Comment (dhudsmith) [inline utils/tweet_search.py:219]: isn't this duplicated from the user code (and probably the search as well)?

File: utils/user.py [hunk 5]
```diff
@@ -29,12 +29,13 @@ def __init__(self,
         self.query_params = query_params
         self.ident_type = ident_type
 
-    def pull(self, ident: Union[List[str], str], output_dir: str, batch_size: int = 100):
+    def pull(self, ident: Union[List[str], str], output_dir: str, save_format: str = 'csv', batch_size: int = 100):
```
Comment (dhudsmith) [inline utils/user.py:32]: use class members instead of args?

**Review by dhudsmith — CHANGES_REQUESTED** [review]
Very nice, Nick! You really went the extra mile here hitting lots of requested features. I tested and things are working great so far. I read through the code changes and left some minor comments and questions. Let me know what you think!

File: Readme.md [hunk 6]
```diff
@@ -33,53 +33,65 @@ python pull_twitter.py --config_file <path to config yaml file> <subcommand> <su
...
+| --author-id-column | -aic | Name of handles column in handles-csv. Incompatible with handle-column. | No (mutually exclusive with above) | "author_id" |
+| --skip-column | -sc | Name of column containing skip indicators in handles-csv (skip indicated with a 1) | No | "skip" |
+| --use-skip | -usc | Indicates whether to use the skip column to ignore specific handles | No |  |
 
 ### Example
 ```python pull_twitter.py --config-file ./configs/config.yaml timeline -hi "./data/celeb_handle_test.csv" -oh True```
 
 ## Fetch User Data
 
 Using the subcommand `users` will collect profile information connected to each non-skipped user as indicated by the handles_csv parameter.
```
[the remark this answers is not in the document]
Comment (NickDeas) [reply Readme.md:65]: Corrected with recent round of commits

File: utils/timeline.py [hunk 7]
```diff
@@ -86,14 +90,84 @@ def pull(self,
 
             # insert tweets into file
             tweets: List[dict] = response.data
+            has_refs: bool = 'referenced_tweets' in self.query_params.tweet_fields
+            inc_tweets: List[dict] = response.includes['tweets'] if 'tweets' in response.includes.keys() else None
+            inc_users: List[dict] = response.includes['users'] if 'users' in response.includes.keys() else None
+            # inc_media: List[dict] = response.includes['media'] if 'media' in response.includes.keys() else None
```
[the remark this answers is not in the document]
Comment (NickDeas) [reply utils/timeline.py:96]: Added a TODO message

File: utils/timeline.py (hunk 2 above)
Comment (NickDeas) [reply utils/timeline.py:115]: Attempted to pass dtype=str to the constructor, but in the pandas code, it seems like the values are coerced to floats before being forced to strings with this parameter, so the decimal points and scientific notation still remain with this approach.

File: Readme.md [hunk 8]
```diff
@@ -33,53 +33,65 @@ python pull_twitter.py --config_file <path to config yaml file> <subcommand> <su
 
 All data will be saved to the directory indicated by output_dir in the designated config file.  Each subcommand is provided an independent subdirectory to save outputs, and all results are stored in timestamped directories within.
 
+If expansions are designated in the config file, additional output csv's are created to hold the additional data:
+* author_id
+	* for tweet-based outputs, this expansion creates a data file "data_authors.csv" holding user data
```
[the remark this answers is not in the document]
Comment (NickDeas) [reply Readme.md]: Output csv renamed

File: utils/timeline.py (hunk 2 above)
Comment (NickDeas) [reply utils/timeline.py]: fix_floats removed

File: utils/tweet_search.py (hunk 4 above)
Comment (NickDeas) [reply utils/tweet_search.py:218]: fix_floats removed

**Pull request merged**